### PR TITLE
[server][controller] Add PubSubPosition interface

### PR DESCRIPTION
**Pull request by sushantmane**
## Add PubSubPosition interface
Introduced an interface called `PubSubPosition` which will enable us to use positions that have pub-sub-specific    
details, unlike the `long` that we use currently. Additionally, we have introduced `PubSubPositionWireFormat` which    
should be utilized when sending positions over the wire or storing them on disk.

## How was this PR tested?

GHCI

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionType.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.linkedin.venice.pubsub;
+
+// Do not change the values of these constants.
+public class PubSubPositionType {
+  public static final int APACHE_KAFKA_OFFSET = 0;
+  public static final int LI_INTERNAL_XC = 1;
+  public static final int LI_INTERNAL_RESERVED = 2;
+}
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionType.java:8]: Hmm... this doesn't seem right. If there is a need for such mapping, and the mapping needs to support internal extensions, then we shouldn't let it be hardcoded. I can see two options which seem clean:

1. The mapping is fully dynamic, e.g. coming from config or perhaps from the controller (exposed via endpoint, via some state in ZK, or whatever).
2. The mapping is hard-coded for pub subs which are available in open source, e.g. starting from zero and going up from there, and furthermore there is an ability to inject internal mappings via whichever mechanism is chosen to achieve (1), e.g. occupying negative numbers.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionUtils.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+   * @return the position object
+   */
+  public static PubSubPosition fromBytes(byte[] bytes) {
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    int positionType = buffer.getInt();
+    if (positionType == PubSubPositionType.APACHE_KAFKA_OFFSET) {
+      return new ApacheKafkaOffsetPosition(buffer.getLong());
+    }
+    throw new IllegalArgumentException("Unknown position type: " + positionType);
+  }
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionUtils.java:24]: Let's have a proper Avro schema registered in `AvroProtocolDefinition` for this and leveraging `InternalAvroSpecificSerializer` for serde and versioning. Let's not cook up some hand-coded protocol please.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubPosition.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import com.linkedin.venice.pubsub.PubSubPositionUtils;
+
+
+/**
+ * Represents a position of a message in a partition of a topic.
+ */
+public interface PubSubPosition {
+  int comparePosition(PubSubPosition other);
+
+  long diff(PubSubPosition other);
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubPosition.java:12]: Is this only going to be supportable in Kafka? Is it viable to have this in the API?

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionType.java (hunk 1 above)
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionType.java:8]: > If there is a need for such mapping
I thinking if we can use magic byte to achieve the same effect

Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionType.java]: As per our discussion, using the `type` field in `PubSubPositionWireFormat` to include the above info.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionUtils.java (hunk 2 above)
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionUtils.java]: Done.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/kafka/ApacheKafkaOffsetPosition.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+public class ApacheKafkaOffsetPosition implements PubSubPosition {
+  private final long offset;
+
+  public ApacheKafkaOffsetPosition(long offset) {
+    this.offset = offset;
+  }
+
+  public ApacheKafkaOffsetPosition(ByteBuffer buffer) {
+    // read the first 8 bytes as a long
+    this(Objects.requireNonNull(buffer, "Cannot create ApacheKafkaOffsetPosition with null").getLong(0));
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/kafka/ApacheKafkaOffsetPosition.java]: Done.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+   * Converts a wire format position to a PubSubPosition
+   * @param positionWireFormat the wire format position
+   * @return concrete position object represented by the wire format
+   */
+  PubSubPosition convertToPosition(PubSubPositionWireFormat positionWireFormat) {
+    if (positionWireFormat == null) {
+      throw new IllegalArgumentException("Cannot deserialize null wire format position");
+    }
+
+    if (positionWireFormat.type == PubSubPositionType.APACHE_KAFKA_OFFSET) {
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java:28]: Nitpick: I assume when we add more open source pub sub implementations, this will eventually become a switch/case. Shall we make it a switch to begin with?

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubPosition.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package com.linkedin.venice.pubsub.api;
+
+import com.linkedin.venice.pubsub.PubSubPositionFactory;
+
+
+/**
+ * Represents a position of a message in a partition of a topic.
+ */
+public interface PubSubPosition {
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubPosition.java:9]: Is the intent that eventually, `PubSubMessage<K, V, OFFSET>` will become `PubSubMessage<K, V, OFFSET extends PubSubPosition>`? It's ok to do it separately in another commit, but just looking to see IIUC.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java (hunk 5 above)
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java:28]: Yeah, will replace `if` with `switch`.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubPosition.java (hunk 6 above)
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubPosition.java:9]: Since we're using the interface for representing the position I believe we do not need to have `OFFSET` in `PubSubMessage<K, V, OFFSET>`. In fact, I would like to get rid of `K` and `V` as well from `PubSubMessage`.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/kafka/ApacheKafkaOffsetPosition.java [hunk 7]
```diff
@@ -98,8 +111,14 @@ public int hashCode() {
   public PubSubPositionWireFormat getPositionWireFormat() {
     PubSubPositionWireFormat wireFormat = new PubSubPositionWireFormat();
     wireFormat.type = PubSubPositionType.APACHE_KAFKA_OFFSET;
-    wireFormat.rawBytes = (ByteBuffer) ByteBuffer.allocate(Long.BYTES).putLong(offset).flip(); // flip to avoid buffer
-                                                                                               // underflow
+    // write the offset as avro serialized long
+    try {
+      ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/kafka/ApacheKafkaOffsetPosition.java]: Done

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java (hunk 5 above)
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java]: Done

File: internal/venice-client-common/src/main/java/com/linkedin/venice/io/ZeroCopyByteArrayOutputStream.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  public ZeroCopyByteArrayOutputStream(int size) {
+    super(size);
+  }
+
+  @Override
+  public synchronized byte[] toByteArray() {
+    return buf;
+  }
+
+  public synchronized ByteBuffer toByteBuffer() {
```
Comment (FelixGV) [inline internal/venice-client-common/src/main/java/com/linkedin/venice/io/ZeroCopyByteArrayOutputStream.java:17]: Should we use this convenience method in `GzipCompressor::compress` since it's doing basically the same thing?

**Review by FelixGV — DISMISSED** [review]
LGTM, though you seem to have an error in static analysis? Please take a look.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+   * @param positionWireFormat the wire format position
+   * @return concrete position object represented by the wire format
+   */
+  PubSubPosition convertToPosition(PubSubPositionWireFormat positionWireFormat) {
+    if (positionWireFormat == null) {
+      throw new IllegalArgumentException("Cannot deserialize null wire format position");
+    }
+
+    switch (positionWireFormat.type) {
+      case PubSubPositionType.APACHE_KAFKA_OFFSET:
```
Comment (ZacAttack) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java:29]: Maybe worth noting with either javadoc or code comment, but the implication here is that all subscribers to a topic will need to be at some X version of this object before it can be converted.  Otherwise conversion will trigger exceptions.  When thinking about Venice core components, maybe not a bit deal, but since this will go into clients, it implies we'll need to be able to track those clients and get them to upgrade.

Comment (FelixGV) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java:29]: Hmm, but what is the alternative? What could we do besides throwing if non-upgraded code encounters a type it cannot deal with?

Comment (ZacAttack) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java:29]: Been meaning to have as an inday project to have our jars signed with the release number and then emit that jar number to metrics so we can track this stuff better.  Not relevant to this rb, but something to consider if we're worried about tracking this stuff as it is deployed.

File: internal/venice-client-common/src/main/java/com/linkedin/venice/io/ZeroCopyByteArrayOutputStream.java (hunk 8 above)
Comment (sushantmane) [reply internal/venice-client-common/src/main/java/com/linkedin/venice/io/ZeroCopyByteArrayOutputStream.java:17]: Good catch, thanks! Updated PR to use `ZCBYOS::toByteBuffer` in `GzipCompressor::compress`

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java (hunk 9 above)
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java:29]: @ZacAttack - I'm not sure if I fully understood your point. Here is what I think:
1. To begin with `PubSubPosition` should only be used by brand-new clients such as CC.
  As all clients will have the type of offset from the beginning, there won't be any issue

2. For the existing clients we need to support (initially) `long` to `PubSubPosition`(`ApacheKafkaOffsetPosition`)  conversion and vice versa.  This is not supported by the code in this PR. We'll create it in a separate PR.  Till then we will continue using `long`. 
 If we do not use `PubSubPosition` in such cases shouldn't be any issue.

**Review by FelixGV — APPROVED** [review]
Thanks for the extra tests 👌 !!

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java (hunk 9 above)
Comment (ZacAttack) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java:29]: I don't have an alternative, I think this is the right thing to do.  I was just noting this property and suggesting a comment somewhere.

**Review by ZacAttack — APPROVED** [review]

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java (hunk 9 above)
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java:29]: Ack.  The comment you suggested should be added to the `PubSubPositionFactory` extension IMO. Let's tackle it later. Thanks!

**Comment by sushantmane** [comment]
Thanks a lot, @FelixGV & @ZacAttack!
